[media controls] overflow button appears clipped on Google Research page

graouts opened the pull request:
#### 2972c6591e214a15e9bec189e9c679ff44bbc8c5
<pre>
[media controls] overflow button appears clipped on Google Research page
<a href="https://bugs.webkit.org/show_bug.cgi?id=254165">https://bugs.webkit.org/show_bug.cgi?id=254165</a>
rdar://104744690

Reviewed by Devin Rousso.

We must account for _visible_ children when determining whether the right bar container should be shown.

* LayoutTests/media/modern-media-controls/overflow-support/dropped-when-narrow-expected.txt: Added.
* LayoutTests/media/modern-media-controls/overflow-support/dropped-when-narrow.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/inline-media-controls.js:
(InlineMediaControls.prototype.layout):

Canonical link: <a href="https://commits.webkit.org/261912@main">https://commits.webkit.org/261912@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9e1d878c686d7aeee2ef75183bbb04bd642e301a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/113240 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/22373 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1913 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/5015 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121694 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/117373 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23743 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/13546 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/6239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/119027 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/100902 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/106335 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99593 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/1443 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46687 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14672 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/89/builds/1482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/98938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/15382 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20674 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/53473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/17234 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4546 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/inline-media-controls.js:
```diff
@@ -209,7 +209,7 @@ class InlineMediaControls extends MediaControls
 
         const widthLeftOfTimeControl = this.leftContainer.children.length > 0 ? this.leftContainer.width : this.leftContainer.leftMargin;
-        const widthRightOfTimeControl = this.rightContainer.children.length > 0 ? this.rightContainer.width : this.rightContainer.rightMargin;
+        const widthRightOfTimeControl = shownRightContainerButtons.length > 0 ? this.rightContainer.width : this.rightContainer.rightMargin;
```
i think you'll need to reevaluate `shownRightContainerButtons` cause we `button.dropped = true` inside the `for` above which will (eventually) result in `button.visible` being set to `false`

dcrousso commented on LayoutTests/media/modern-media-controls/overflow-support/dropped-when-narrow.html:
```diff
@@ -0,0 +1,33 @@
+
+    shouldBecomeDifferent("shadowRoot.querySelector('button.overflow')", "null", () => {
+        shouldBecomeDifferent("shadowRoot.querySelector('button.overflow').getBoundingClientRect().width", "0", async () => {
+            media.style.width = "275px";
```
this seems oddly specific and potentially fragile if the visual style (e.g. margins) of the media controls changes in the future.  is there any way we can do this that doesnt rely on a hardcoded `width`?

graouts replied:
Yeah, the number could be a different value too, but this is one that corresponds to the current state of how metrics work for media controls _right now_. Alas, I cannot think off the top of my head of a better technique, though I'd be happy to revisit after this initial change is committed. I will however make it a lower value, `200` works fine for instance (still fails prior to this change).

dcrousso commented on Source/WebCore/Modules/modern-media-controls/controls/inline-media-controls.js:
```diff
@@ -208,8 +208,11 @@ class InlineMediaControls extends MediaControls
         this.rightContainer.layout();
 
+        // Update the list of shown buttons in the right container since it may have changed again.
+        shownRightContainerButtons = this.rightContainer.children.filter(button => button.enabled && !button.dropped);
```
NIT: I wonder if we should check for `button.visible` instead here, since that's ultimately what we're really after
```
const visibleRightContainerButtons = this.rightContainer.children.filter(button => button.visible);
```
i think we dont do that above because `dropped` is only checked inside `layout`, which hasn't been called yet

graouts replied:
Excellent point, I'll do this in the commit.

graouts committed:
[media controls] overflow button appears clipped on Google Research page
https://bugs.webkit.org/show_bug.cgi?id=254165
rdar://104744690

Reviewed by Devin Rousso.

We must account for _visible_ children when determining whether the right bar container should be shown.

* LayoutTests/media/modern-media-controls/overflow-support/dropped-when-narrow-expected.txt: Added.
* LayoutTests/media/modern-media-controls/overflow-support/dropped-when-narrow.html: Added.
* Source/WebCore/Modules/modern-media-controls/controls/inline-media-controls.js:
(InlineMediaControls.prototype.layout):

Canonical link: https://commits.webkit.org/261912@main
